use tracked remote for C status flag

tkruse opened the pull request:
Do not merge yet, this contains the commits of #98 plus the controversial changes discussed at 
https://groups.google.com/forum/#!topic/ros-sig-buildsystem/wLINOnzjLpc

tkruse commented:
So, @wkentaro, I think to change this commit of yours be okay with the discussion in the google group, it needs a change. First we need to check whether origin/branch exists, and only if it does not, try the tracking branch.

wkentaro commented:
so, all necessary functions have been implemented, haven't it?
the logic in wstool will be like:
```
parent_branch = get_parent_branch(current_branch, allow_other_remote=True)
if is_remote_branch(current_branch, 'origin'):
    if get_version('origin/'+current_branch) == get_version(curr_branch):
        mflag += C
elif: parent_branch is not None:
    if get_version(parent_branch) == get_version(curr_branch):
        mflag += C

tkruse commented:
You are right, that would work. So @wjwwood this is ready for your review and merge. Review could mostly assure no other function is changed. is_remote_branch is an "internal" method.

tkruse commented:
Hm, actually it would not be good for wstool to use is_remote_branch() as suggested by @wkentaro in the comment above, because this function is an internal API. I think the logic must go into get_remote_version().
The git code around get_parent_branch becomes quite ugly, so I refactored it and squashed. The result is not exactly as suggested in the google group, because checking origin/<branchname> as default is ugly (and we never did before).

Up to now, we used the tracking branch, unless it was not "origin", in which case we did nothing. So the contract never was to check on "origin" for a branch/reference of the same name.

With the current change, if the tracked remote is something else, we use that. Overall this is somewhat safer than the change suggested earlier (checking 'origin/<refname>'), as there might be weird name clashes that would be more confusing than helpful.

tkruse commented:
@wkentaro : if you have the time, please review.

wkentaro commented:
I think your extension of `get_branch_parent` is nice. I like that.
The returning values of `git::get_branch_parent` changed, so should be change other vcses' API as well maybe in the future?

wkentaro commented:
Other than that, the code change looks good to me.

tkruse commented:
get_branch_parent is a git-only internal method. I wonder whether we should rename all those git-internal methods to have a leading underscore.

wkentaro commented:
+1
It helps maintainers to know which function is external api and cannot be changed without care.

tkruse commented:
Maybe the rename of internal methods can be a separate PR

wkentaro commented:
> Maybe the rename of internal methods can be a separate PR

I agree that.

wkentaro commented:
So, can this PR be merged?

tkruse commented:
@wjwwood : plz review and merge when you have time (ideally before next release)

wjwwood commented on src/vcstools/git.py:
```diff
@@ -283,7 +284,12 @@ def _do_update(self,
-                    branch_parent = self.get_branch_parent(current_branch=current_branch)
+                    (branch_parent, remote) = self.get_branch_parent(current_branch=current_branch)
+                    if remote != 'origin':
+                        # if remote is different, fast_forward would do do wrong
```
This comment doesn't make sense to me. Should it say something like "..., fast forward would be the wrong thing to do"?

wjwwood commented on src/vcstools/git.py:
```diff
@@ -325,7 +331,12 @@ def _do_update(self,
-                new_branch_parent = self.get_branch_parent(current_branch=refname)
+                (new_branch_parent, remote) = self.get_branch_parent(current_branch=refname)
+                if remote != 'origin':
+                    # if remote is different, fast_forward would do do wrong
```
Same comment.

wjwwood commented on src/vcstools/git.py:
```diff
@@ -335,22 +346,28 @@ def _do_update(self,
-                remote_branch = remote_branch[len('origin/'):]
+            # 'branchname' is shorthand for 'origin/branchname'
+            if remote != 'origin':
+                remote_branch = remote + '/' + remote_branch
```
Would make sense to always prepend `<remote>/`? Or is the shorthand for more readability in the output? Technically speaking, `<branch>` and `origin/<branch>` are different things, in `git` parlance, since `<branch>` is the local version and `origin/<branch>` is specifically what we expect the remote's branch state is like. Maybe that doesn't matter here though.

tkruse replied:
I can rewrite the conditions a bit. We will display something like "foo < bar" in case foo is tracking origin/bar. I guess foo might also track a local branch bar, but that seems so weird that I don't think it is relevant enough.

wjwwood commented on src/vcstools/git.py:
```diff
@@ -473,13 +490,16 @@ def get_status(self, basepath=None, untracked=False):
         checks list of remote branches for match. Set fetch to False if you just fetched already.
 
-        :returns: True if git branch knows ref for remote "origin"
+        :returns: True if git branch knows ref for remote_name
```
... or "origin" if remote_name is None.

wjwwood commented on src/vcstools/git.py:
```diff
@@ -473,13 +490,16 @@ def get_status(self, basepath=None, untracked=False):
         return response
 
-    def is_remote_branch(self, branch_name, fetch=True):
+    def is_remote_branch(self, branch_name, remote_name=None, fetch=True):
```
Should `remote_name` just default to `"origin"`?

tkruse replied:
I missed that. It could also do, but if user passes None, the function should still use 'origin' not sure if that's how python behaves...

wjwwood replied:
I didn't think about the user passing `None` for origin. This is fine as is then.

wjwwood commented:
Other than a few comments, lgtm. I'll merge after the comments are addressed or otherwise resolved.

tkruse commented:
fixed as per comments, squashed, rebased
